src/ctx.rs

```diff
@@ -23,54 +78,85 @@ struct PersistedVariable {
     default_value: serde_yaml::Value,
 }
 
-impl TryFrom<PersistedOptions> for Variables {
+impl TryFrom<Vec<PersistedVariable>> for Variables {
     type Error = crate::Error;
-    fn try_from(persisted: PersistedOptions) -> Result<Self> {
+    fn try_from(persisted: Vec<PersistedVariable>) -> Result<Self> {
         let mut out = Variables::default();
-        for saved_var in persisted.variables {
+        for saved_var in persisted {
             out.insert(saved_var.name, saved_var.default_value)?;
         }
         Ok(out)
     }
 }
 
-impl From<Variables> for PersistedOptions {
+impl From<Variables> for Vec<PersistedVariable> {
     fn from(variables: Variables) -> Self {
-        let formatted_variables = variables
+        variables
             .tree()
             .iter()
             .map(|(k, v)| PersistedVariable {
                 name: k.into(),
                 default_value: v.clone(),
             })
-            .collect::<Vec<PersistedVariable>>();
-        PersistedOptions {
-            variables: formatted_variables,
-        }
+            .collect::<Vec<PersistedVariable>>()
     }
 }
 
-pub fn extract_variables(ctx: &Ctx) -> Result<(Variables, Variables)> {
-    let mut confirmed_variables = Variables::default();
-    confirmed_variables.insert(
+pub(crate) fn extract_variables(ctx: &Ctx) -> Result<(Variables, Variables, Variables)> {
```

Comment:
function should return a struct at least to document what is what instead of this tuple of items with the same type.

```
pub(crate) struct VariablesFromCtx{
  src: Variables,
  cli: Variables,
  saved: Variables,
}
```